openwisp_firmware_upgrader/static/firmware-upgrader/js/upgrade-progress.js

```diff
@@ -0,0 +1,494 @@
+  if (shouldScroll) {
+    scrollToBottom(logElement);
+  }
+
+  // Update modified timestamp
+  if (operation.modified) {
+    operationFieldset
+      .find(".field-modified .readonly")
+      .html(getFormattedDateTimeString(operation.modified));
+  }
+
+  let colorCode = getStatusColor(operation.status);
+  operationFieldset.css("background-color", colorCode);
+  setTimeout(function () {
+    operationFieldset.addClass("object-updated");
+    operationFieldset.css("background-color", "inherit");
+  }, 100);
+}
+
+function updateStatusWithProgressBar(statusField, operation) {
+  let $ = django.jQuery;
+
+  let status = operation.status;
+  // Get log content to calculate real progress
+  let logContent = operation.log || "";
+  let progressPercentage = getProgressPercentage(status, logContent);
+  let progressClass = status.replace(/\s+/g, "-");
+
+  if (!statusField.find(".upgrade-status-container").length) {
+    statusField.empty();
+    statusField.append('<div class="upgrade-status-container"></div>');
+  }
+
+  let statusContainer = statusField.find(".upgrade-status-container");
+
+  let statusHtml = `
+    <span class="upgrade-status-${progressClass}">${status}</span>
+  `;
+
+  // Add progress bar for all statuses
+  if (status === "in-progress" || status === "in progress") {
+    statusHtml += `
+      <div class="upgrade-progress-bar">
+        <div class="upgrade-progress-fill in-progress" style="width: ${progressPercentage}%"></div>
+      </div>
+      <span class="upgrade-progress-text">${progressPercentage}%</span>
+    `;
+  } else if (status === "success") {
+    statusHtml += `
+      <div class="upgrade-progress-bar">
+        <div class="upgrade-progress-fill success" style="width: 100%"></div>
+      </div>
+      <span class="upgrade-progress-text">100%</span>
+    `;
+  } else if (status === "failed" || status === "aborted") {
+    statusHtml += `
+      <div class="upgrade-progress-bar">
+        <div class="upgrade-progress-fill ${status}" style="width: ${progressPercentage}%"></div>
+      </div>
+      <span class="upgrade-progress-text">${progressPercentage}%</span>
+    `;
+  } else {
+    statusHtml += `
+      <div class="upgrade-progress-bar">
+        <div class="upgrade-progress-fill" style="width: ${progressPercentage}%"></div>
+      </div>
+      <span class="upgrade-progress-text">${progressPercentage}%</span>
+    `;
+  }
+
+  statusContainer.html(statusHtml);
+}
+
+function getProgressPercentage(status, logContent = "") {
+  if (status === "success") {
+    return 100;
+  } else if (status === "failed" || status === "aborted") {
+    return calculateProgressFromLog(logContent);
+  } else if (status === "in-progress" || status === "in progress") {
+    return calculateProgressFromLog(logContent);
+  }
+  return 0;
+}
+
+function calculateProgressFromLog(logContent = "") {
+  if (!logContent) return 0;
+
+  const upgradeSteps = [
+    { keyword: "Connection successful, starting upgrade", progress: 10 },
```

Comment:
This is very fragile! :x: :warning: 
What happens if the UI is translated to other languages?

@Aryamanz29 @okraits watch out!